### PR TITLE
[TACHYON-1141] Update AbstractTFS & company to use TachyonFileSystem

### PR DESCRIPTION
**Pull request by aaudiber**
https://tachyon.atlassian.net/browse/TACHYON-1141

### CODE REVIEW

File: clients/unshaded/src/main/java/tachyon/client/file/FileSystemContext.java [hunk 1]
```diff
@@ -72,4 +72,8 @@ public synchronized void reset() {
     mFileSystemMasterClientPool =
         new FileSystemMasterClientPool(ClientContext.getMasterAddress());
   }
+
+  public static FileSystemContext get() {
```
Comment (aaudiber) [inline clients/unshaded/src/main/java/tachyon/client/file/FileSystemContext.java]: Added to make it easier to mock places that would use FileSystemContext.INSTANCE. They can instead use FileSystemContext.get(), which can be mocked.

File: clients/unshaded/src/main/java/tachyon/hadoop/AbstractTFS.java [hunk 2]
```diff
@@ -367,19 +395,25 @@ public void initialize(URI uri, Configuration conf) throws IOException {
       mStatistics.incrementReadOps(1);
     }
 
-    if (!mTFS.exist(tPath)) {
+    List<FileInfo> files;
+    try {
+      TachyonFile file = mTFS.open(tPath);
+      files = mTFS.listStatus(file);
+    } catch (InvalidPathException e) {
       throw new FileNotFoundException("File does not exist: " + path);
```
Comment (aaudiber) [inline clients/unshaded/src/main/java/tachyon/hadoop/AbstractTFS.java]: Are we committed to moving to Java 7? If so we can use (InvalidPathException | FileDoesNotExistException e) instead

File: clients/unshaded/src/main/java/tachyon/hadoop/AbstractTFS.java [hunk 3]
```diff
@@ -58,13 +67,15 @@
   public static final String RECOMPUTE_PATH = "tachyon_recompute/";
 
   private static final Logger LOG = LoggerFactory.getLogger(Constants.LOGGER_TYPE);
+  // Always tell Hadoop that we have 3x replication.
+  private static final int BLOCK_REPLICATION_CONSTANT = 3;
```
Comment (yupeng9) [inline clients/unshaded/src/main/java/tachyon/hadoop/AbstractTFS.java]: Shall we make this configurable by the user?

File: clients/unshaded/src/main/java/tachyon/hadoop/AbstractTFS.java [hunk 4]
```diff
@@ -190,14 +220,19 @@ public boolean delete(Path cPath, boolean recursive) throws IOException {
...
+    DeleteOptions options =
+        new DeleteOptions.Builder(ClientContext.getConf()).setRecursive(recursive).build();
+    try {
+      TachyonFile file = mTFS.open(path);
+      mTFS.delete(file, options);
+      return true;
+    } catch (InvalidPathException e) {
       return false;
+    } catch (FileDoesNotExistException e) {
+      return false;
```
Comment (yupeng9) [inline clients/unshaded/src/main/java/tachyon/hadoop/AbstractTFS.java]: Shall we log some message?

File: clients/unshaded/src/main/java/tachyon/hadoop/HdfsFileInputStream.java [hunk 5]
```diff
@@ -73,7 +77,7 @@
    * @param tachyonConf Tachyon configuration
    * @throws IOException if the underlying file does not exist or its stream cannot be created
    */
-  public HdfsFileInputStream(TachyonFS tfs, long fileId, Path hdfsPath, Configuration conf,
+  public HdfsFileInputStream(TachyonFileSystem tfs, long fileId, Path hdfsPath, Configuration conf,
```
Comment (yupeng9) [inline clients/unshaded/src/main/java/tachyon/hadoop/HdfsFileInputStream.java]: Is this a public API?

Comment (yupeng9) [inline clients/unshaded/src/main/java/tachyon/hadoop/HdfsFileInputStream.java]: Also, do we need to pass `TachyonFileSystem` in? Can we get it from `TachyonFileSystemFactory`?

Comment (yupeng9) [inline clients/unshaded/src/main/java/tachyon/hadoop/HdfsFileInputStream.java]: And likewise for TachyonConf

**Comment by yupeng9** [comment]
Thanks for migrating the implementation of the old API. I left some comments.

File: clients/unshaded/src/main/java/tachyon/hadoop/AbstractTFS.java (hunk 3 above)
Comment (aaudiber) [inline clients/unshaded/src/main/java/tachyon/hadoop/AbstractTFS.java]: I think this is how many replicas to tell HDFS that we currently have, not how many we want to have. The functionality hasn't changed, we're just using a constant now instead of the previous
```
  public int getDiskReplication() {
    // TODO(hy): Implement it.
    return 3;
  }
```
We can decide how to manage block replication info separately from this commit.

File: clients/unshaded/src/main/java/tachyon/hadoop/AbstractTFS.java (hunk 4 above)
Comment (aaudiber) [inline clients/unshaded/src/main/java/tachyon/hadoop/AbstractTFS.java]: Done

File: clients/unshaded/src/main/java/tachyon/hadoop/HdfsFileInputStream.java (hunk 5 above)
Comment (aaudiber) [inline clients/unshaded/src/main/java/tachyon/hadoop/HdfsFileInputStream.java]: @calvinjia Is this public api?

@yupeng9 Yep, we could remove both of those parameters and use the singletons instead.

Comment (calvinjia) [inline clients/unshaded/src/main/java/tachyon/hadoop/HdfsFileInputStream.java]: Users shouldn't directly use `HdfsFileInputStream`, it is wrapped in Hadoop's `FSDataInputStream`

**Comment by yupeng9** [comment]
LGTM

File: clients/unshaded/src/main/java/tachyon/hadoop/HdfsFileInputStream.java [hunk 6]
```diff
@@ -64,36 +69,38 @@
...
+      FileSystem.Statistics stats) throws IOException {
+    LOG.debug("HdfsFileInputStream({}, {}, {}, {}, {})", fileId, hdfsPath, conf,
         bufferSize, stats);
-    mTachyonConf = tachyonConf;
+    mTachyonConf = ClientContext.getConf();
     long bufferBytes = mTachyonConf.getBytes(Constants.USER_FILE_BUFFER_BYTES);
     mBuffer = new byte[Ints.checkedCast(bufferBytes) * 4];
     mCurrentPosition = 0;
-    mTFS = tfs;
+    mTFS = TachyonFileSystem.TachyonFileSystemFactory.get();
```
Comment (calvinjia) [inline clients/unshaded/src/main/java/tachyon/hadoop/HdfsFileInputStream.java]: Can you just call `TachyonFileSystemFactory.get()`?

File: clients/unshaded/src/main/java/tachyon/hadoop/HdfsFileInputStream.java [hunk 7]
```diff
@@ -311,9 +318,9 @@ public void seek(long pos) throws IOException {
     if (pos < 0) {
       throw new IOException("Seek position is negative: " + pos);
     }
-    if (pos > mTachyonFile.length()) {
-      throw new IOException("Seek position is past EOF: " + pos + ", fileSize = "
-          + mTachyonFile.length());
+    if (pos > mFileInfo.getLength()) {
+      throw new IOException(
+          "Seek position is past EOF: " + pos + ", fileSize = " + mFileInfo.getLength());
```
Comment (calvinjia) [inline clients/unshaded/src/main/java/tachyon/hadoop/HdfsFileInputStream.java]: Is there a common error message constant for this one?

File: clients/unshaded/src/main/java/tachyon/hadoop/AbstractTFS.java [hunk 8]
```diff
@@ -76,25 +87,26 @@ public FSDataOutputStream append(Path cPath, int bufferSize, Progressable progre
...
+    } catch (InvalidPathException e) {
+      throw new FileNotFoundException("File does not exist: " + path);
+    } catch (FileDoesNotExistException e) {
+      throw new FileNotFoundException("File does not exist: " + path);
+    } catch (TachyonException e) {
+      throw new IOException(e);
     }
 
-    return new FSDataOutputStream(file.getOutStream(), mStatistics);
+    return new FSDataOutputStream(mTFS.getOutStream(file.getFileId(), OutStreamOptions.defaults()));
```
Comment (calvinjia) [inline clients/unshaded/src/main/java/tachyon/hadoop/AbstractTFS.java]: Missing `mStatistics`?

File: clients/unshaded/src/main/java/tachyon/hadoop/AbstractTFS.java [hunk 9]
```diff
@@ -121,21 +133,41 @@ public FSDataOutputStream create(Path cPath, FsPermission permission, boolean ov
     }
 
     TachyonURI path = new TachyonURI(Utils.getPathWithoutScheme(cPath));
-    if (mTFS.exist(path)) {
-      if (overwrite && !mTFS.getFileStatus(-1, path).isFolder) {
-        if (!mTFS.delete(path, false)) {
-          throw new IOException("Failed to delete existing data " + cPath);
+    try {
+      TachyonFile file = mTFS.open(path);
```
Comment (calvinjia) [inline clients/unshaded/src/main/java/tachyon/hadoop/AbstractTFS.java]: I think you want to use `openIfExists`?

File: clients/unshaded/src/main/java/tachyon/hadoop/AbstractTFS.java [hunk 10]
```diff
@@ -121,21 +133,41 @@ public FSDataOutputStream create(Path cPath, FsPermission permission, boolean ov
...
+          try {
+            mTFS.delete(file);
+          } catch (TachyonException e) {
+            throw new IOException("Failed to delete existing data " + cPath, e);
+          }
         }
       } else {
-        throw new IOException(cPath.toString() + " already exists. Directories cannot be "
-            + "overwritten with create.");
+        throw new IOException(cPath.toString() + " already exists.");
```
Comment (calvinjia) [inline clients/unshaded/src/main/java/tachyon/hadoop/AbstractTFS.java]: Do we have any constant error messages for these cases?

File: clients/unshaded/src/main/java/tachyon/hadoop/AbstractTFS.java [hunk 11]
```diff
@@ -121,21 +133,41 @@ public FSDataOutputStream create(Path cPath, FsPermission permission, boolean ov
...
       }
+    } catch (InvalidPathException e) {
+      // The file doesn't exist, fall through to create it.
```
Comment (calvinjia) [inline clients/unshaded/src/main/java/tachyon/hadoop/AbstractTFS.java]: I don't think you want to fall through here?

File: clients/unshaded/src/main/java/tachyon/hadoop/AbstractTFS.java [hunk 12]
```diff
@@ -121,21 +133,41 @@ public FSDataOutputStream create(Path cPath, FsPermission permission, boolean ov
...
+      outStream = mTFS.getOutStream(path, options);
+    } catch (TachyonException e) {
+      throw new IOException(e);
     }
-    long fileId = mTFS.createFile(path, blockSize);
-    TachyonFile file = mTFS.getFile(fileId);
-    file.setUFSConf(getConf());
 
-    return new FSDataOutputStream(file.getOutStream(), mStatistics);
+    return new FSDataOutputStream(outStream, mStatistics);
```
Comment (calvinjia) [inline clients/unshaded/src/main/java/tachyon/hadoop/AbstractTFS.java]: Perhaps you can just do this in the `try catch`? I notice you use a different style in the `delete` method (return from the try catch)

File: clients/unshaded/src/main/java/tachyon/hadoop/AbstractTFS.java [hunk 13]
```diff
@@ -461,4 +511,42 @@ public void setWorkingDirectory(Path path) {
...
+
+  /**
+   * Opens a {@link TachyonFile} for the given path, throwing FileNotFoundException with the
+   * specified error message if the path doesn't exist.
+   *
+   * @param tPath the path to look up
+   * @param errorMessage the message for the FileNotFoundException thrown if the path doesn't exist
+   * @throws IOException if an unexpeced Tachyon exception occurs
+   */
+  private TachyonFile tryOpen(TachyonURI tPath, String errorMessage) throws IOException {
```
Comment (calvinjia) [inline clients/unshaded/src/main/java/tachyon/hadoop/AbstractTFS.java]: Why is this necessary?

File: clients/unshaded/src/main/java/tachyon/hadoop/AbstractTFS.java [hunk 14]
```diff
@@ -270,25 +304,30 @@ public FileStatus getFileStatus(Path path) throws IOException {
...
     }
-    TachyonFile file;
+    FileInfo fileStatus;
     try {
-      file = mTFS.getFile(tPath);
-    } catch (IOException ioe) {
+      TachyonFile file = mTFS.open(tPath);
+      fileStatus = mTFS.getInfo(file);
+    } catch (InvalidPathException e) {
+      LOG.info("File does not exist: " + path);
```
Comment (calvinjia) [inline clients/unshaded/src/main/java/tachyon/hadoop/AbstractTFS.java]: Should this log message be more specific?

File: clients/unshaded/src/main/java/tachyon/hadoop/AbstractTFS.java (hunk 2 above)
Comment (calvinjia) [inline clients/unshaded/src/main/java/tachyon/hadoop/AbstractTFS.java]: Let's not do that until we really move over.

File: clients/unshaded/src/main/java/tachyon/hadoop/AbstractTFS.java (hunk 8 above)
Comment (aaudiber) [inline clients/unshaded/src/main/java/tachyon/hadoop/AbstractTFS.java]: Good catch, done

File: clients/unshaded/src/main/java/tachyon/hadoop/AbstractTFS.java (hunk 9 above)
Comment (aaudiber) [inline clients/unshaded/src/main/java/tachyon/hadoop/AbstractTFS.java]: Done

File: clients/unshaded/src/main/java/tachyon/hadoop/AbstractTFS.java (hunk 10 above)
Comment (aaudiber) [inline clients/unshaded/src/main/java/tachyon/hadoop/AbstractTFS.java]: Done

File: clients/unshaded/src/main/java/tachyon/hadoop/AbstractTFS.java (hunk 11 above)
Comment (aaudiber) [inline clients/unshaded/src/main/java/tachyon/hadoop/AbstractTFS.java]: Done

File: clients/unshaded/src/main/java/tachyon/hadoop/AbstractTFS.java (hunk 12 above)
Comment (aaudiber) [inline clients/unshaded/src/main/java/tachyon/hadoop/AbstractTFS.java]: Done

File: clients/unshaded/src/main/java/tachyon/hadoop/AbstractTFS.java (hunk 14 above)
Comment (aaudiber) [inline clients/unshaded/src/main/java/tachyon/hadoop/AbstractTFS.java]: Done

File: clients/unshaded/src/main/java/tachyon/hadoop/AbstractTFS.java (hunk 13 above)
Comment (aaudiber) [inline clients/unshaded/src/main/java/tachyon/hadoop/AbstractTFS.java]: It's just a convenience method for places where we want to avoid try/catch.

File: clients/unshaded/src/main/java/tachyon/hadoop/HdfsFileInputStream.java (hunk 6 above)
Comment (aaudiber) [inline clients/unshaded/src/main/java/tachyon/hadoop/HdfsFileInputStream.java]: Done

File: clients/unshaded/src/main/java/tachyon/hadoop/HdfsFileInputStream.java (hunk 7 above)
Comment (aaudiber) [inline clients/unshaded/src/main/java/tachyon/hadoop/HdfsFileInputStream.java]: Done

File: clients/unshaded/src/main/java/tachyon/client/file/FileSystemContext.java (hunk 1 above)
Comment (calvinjia) [inline clients/unshaded/src/main/java/tachyon/client/file/FileSystemContext.java]: Is there any way to enforce callers use `FileSystemContext.get()` instead of `FileSystemContext.INSTANCE`?

**Comment by aaudiber** [comment]
Jenkins, test this please.

File: clients/unshaded/src/main/java/tachyon/client/file/FileSystemContext.java (hunk 1 above)
Comment (aaudiber) [inline clients/unshaded/src/main/java/tachyon/client/file/FileSystemContext.java]: I've switched to mocking `FileSystemContext.INSTANCE` directly with Whitebox.

File: clients/unshaded/src/main/java/tachyon/hadoop/AbstractTFS.java [hunk 15]
```diff
@@ -121,21 +132,35 @@ public FSDataOutputStream create(Path cPath, FsPermission permission, boolean ov
...
       }
+    } catch (TachyonException e) {
+      throw new IOException(e);
     }
-    long fileId = mTFS.createFile(path, blockSize);
-    TachyonFile file = mTFS.getFile(fileId);
-    file.setUFSConf(getConf());
 
-    return new FSDataOutputStream(file.getOutStream(), mStatistics);
+    OutStreamOptions options =
```
Comment (calvinjia) [inline clients/unshaded/src/main/java/tachyon/hadoop/AbstractTFS.java]: Please add a comment like `// file does not exist` before this block

File: clients/unshaded/src/main/java/tachyon/hadoop/AbstractTFS.java [hunk 16]
```diff
@@ -121,21 +132,35 @@ public FSDataOutputStream create(Path cPath, FsPermission permission, boolean ov
...
+        }
+        FileInfo info = mTFS.getInfo(file);
+        if (info.isIsFolder()) {
+          throw new IOException(
+              ExceptionMessage.FILE_CREATE_IS_DIRECTORY.getMessage(cPath.toString()));
+        }
+        try {
+          mTFS.delete(file);
+        } catch (TachyonException e) {
+          throw new IOException(ExceptionMessage.DELETE_FAILED.getMessage(cPath), e);
```
Comment (calvinjia) [inline clients/unshaded/src/main/java/tachyon/hadoop/AbstractTFS.java]: Why do we make this exception more explicit?

File: clients/unshaded/src/main/java/tachyon/hadoop/AbstractTFS.java [hunk 17]
```diff
@@ -160,11 +185,9 @@ public FSDataOutputStream create(Path cPath, FsPermission permission, boolean ov
...
   public FSDataOutputStream createNonRecursive(Path cPath, FsPermission permission,
       boolean overwrite, int bufferSize, short replication, long blockSize, Progressable progress)
-      throws IOException {
-    TachyonURI path = new TachyonURI(Utils.getPathWithoutScheme(cPath.getParent()));
-    if (!mTFS.exist(path)) {
-      throw new FileNotFoundException("Parent directory does not exist!");
-    }
+          throws IOException {
+    TachyonURI parentPath = new TachyonURI(Utils.getPathWithoutScheme(cPath.getParent()));
+    tryOpen(parentPath);
```
Comment (calvinjia) [inline clients/unshaded/src/main/java/tachyon/hadoop/AbstractTFS.java]: Is it safe to throw the more general `IOException` here (previous was `FileNotFound`)?

Comment (aaudiber) [inline clients/unshaded/src/main/java/tachyon/hadoop/AbstractTFS.java]: It should be, the Hadoop API says nothing about `FileNotFound`

File: clients/unshaded/src/main/java/tachyon/hadoop/AbstractTFS.java (hunk 16 above)
Comment (aaudiber) [inline clients/unshaded/src/main/java/tachyon/hadoop/AbstractTFS.java]: Changed to just `IOException(e)`

File: clients/unshaded/src/main/java/tachyon/hadoop/AbstractTFS.java (hunk 15 above)
Comment (aaudiber) [inline clients/unshaded/src/main/java/tachyon/hadoop/AbstractTFS.java]: Added comments

**Comment by calvinjia** [comment]
LGTM

File: clients/unshaded/src/main/java/tachyon/hadoop/AbstractTFS.java [hunk 18]
```diff
@@ -461,4 +493,39 @@ public void setWorkingDirectory(Path path) {
...
   }
+
+  /**
+   * Convenience method which opens a {@link TachyonFile} for the given path, wrapping any
+   * {@link TachyonException} in {@link IOException}.
+   *
+   * @param tPath the path to look up
+   * @throws {@link IOException} if a Tachyon exception occurs
+   */
+  private TachyonFile tryOpen(TachyonURI tPath) throws IOException {
```
Comment (gpang) [inline clients/unshaded/src/main/java/tachyon/hadoop/AbstractTFS.java]: NIT: path instead of tPath?

File: common/src/main/java/tachyon/exception/ExceptionMessage.java [hunk 19]
```diff
@@ -54,6 +55,9 @@
   FAILED_SEEK_FORWARD("Failed to seek forward to {0}"),
   FAILED_SKIP("Failed to skip {0}"),
   INSTREAM_CANNOT_SKIP("The underlying BlockInStream could not skip {0}"),
+  READ_CLOSED_STREAM("Cannot read from a closed stream"),
+  SEEK_NEGATIVE("Seek position is negative: {0}"),
+  SEEK_PAST_EOF("Seek position is past EOF: {0}, fileSize = {1}"),
```
Comment (gpang) [inline common/src/main/java/tachyon/exception/ExceptionMessage.java]: NIT: I think "fileSize: ..." would be more consistent than using " = ".

File: common/src/main/java/tachyon/exception/ExceptionMessage.java [hunk 20]
```diff
@@ -83,6 +87,11 @@
   UNEXPECETD_JOURNAL_ENTRY("Unexpected entry in journal: {0}"),
   UNKNOWN_ENTRY_TYPE("Unknown entry type: {0}"),
 
+  // file
+  FILE_ALREADY_EXISTS("{0} already exists"),
+  FILE_CREATE_IS_DIRECTORY("{0} already exists. Directories cannot be overwritten with create"),
+  HDFS_FILE_NOT_FOUND("File {0} with FID {1} is not found."),
```
Comment (gpang) [inline common/src/main/java/tachyon/exception/ExceptionMessage.java]: Unless I am mistaken, I don't think "FID" is a commonly used term in tachyon? I think it should just be "id".

Also, should the trailing period be removed? The other messages don't have the period.

**Comment by gpang** [comment]
I just had some minor comments. Thanks!

File: common/src/main/java/tachyon/exception/ExceptionMessage.java (hunk 20 above)
Comment (aaudiber) [inline common/src/main/java/tachyon/exception/ExceptionMessage.java]: Done.

File: common/src/main/java/tachyon/exception/ExceptionMessage.java (hunk 19 above)
Comment (aaudiber) [inline common/src/main/java/tachyon/exception/ExceptionMessage.java]: Done.

File: clients/unshaded/src/main/java/tachyon/hadoop/AbstractTFS.java (hunk 18 above)
Comment (aaudiber) [inline clients/unshaded/src/main/java/tachyon/hadoop/AbstractTFS.java]: Done.

**Comment by gpang** [comment]
Jenkins, test this please

**Comment by gpang** [comment]
LGTM
